Project Wizard Smoke Tests: Python Project existing interpreter ipykernel tests

sharon-wang opened the pull request:
## Description

- Addresses the Python Project ipykernel tests part of #3879

### Summary
- Reorganizes the automation class for Project Wizard
    - Each step now has a class which is contained in the Project Wizard class
    - Selectors are scoped to the class constructors whenever possible
    - Opting for usage of the `Locator` class when reasonable (sometimes waiting for elements and clicking on them still works better with the BaseElement class)
- Adds a couple ipykernel project wizard tests
- Reorganize Project Wizard smoke tests so all of the Project Wizard tests can be run with one `only`
- Modify the frontend components for the Project Wizard so that wizard steps and substeps have an easily referenceable ID, making constructing selectors easier
    <img width="1470" alt="proj_wiz_set_up_py_env_ids" src="https://github.com/posit-dev/positron/assets/25834218/c6b9028e-3a5a-456f-8062-cc59843fced8">
- Add option to `PositronPythonFixtures` to install ipykernel as part of selecting the interpreter
- Add new method `popupCurrentlyOpen()` to `PositronPopups` to check if a popup is currently open
- Support retrieving the active interpreter path from the Interpreter Dropdown
- Modify `StartInterpreter` methods to support returning the interpreter path
   - NOTE: `StartInterpreter` methods don't seem to be used elsewhere, so we should add some tests to confirm the class methods are working. Possibly in test/smoke/src/areas/positron/top-action-bar/top-action-bar.test.ts?

### 🆕 Smoke Tests
#### Python Project with existing interpreter:
- With ipykernel already installed
- With ipykernel not already installed

## QA Notes

Please add new test IDs and validate on Windows. Thank you :)

sharon-wang committed:
organize new project wizard selectors by modal step

sharon-wang committed:
reorganize project wizard automation class

organizes elements by step in the wizard automation class and encloses all project wizard tests in one `describe` block so they can be run together

sharon-wang committed:
simplify project type button name

sharon-wang committed:
add ids to project wizard steps and substeps

sharon-wang committed:
add support for installing ipykernel if a dialog opens when selecting an interpreter

sharon-wang committed:
add ipykernel install project wizard tests

sharon-wang committed:
format file and clean up comments

sharon-wang committed:
move selector strings to class scope when possible

sharon-wang committed:
apply prettier formatting to positronNewProjectWizard.ts

sharon-wang commented:
I've kicked off a CI build: https://github.com/posit-dev/positron/actions/runs/9876521682

jonvanausdeln committed:
Added TestRail IDs and removed stray `.only`

sharon-wang committed:
don't fail starting the python interpreter if arg not passed

sharon-wang committed:
add support for returning the selected interpreter info

sharon-wang committed:
select interpreter by path instead of version for proj wiz python tests

sharon-wang committed:
add python interpreter selection to `before` for python proj wiz tests

sharon-wang committed:
Revert "add python interpreter selection to `before` for python proj wiz tests"

This reverts commit 5cbca80580ccafd7cec4783f2e2c6e1a9ed1d266.

Add this didn't solve the timeout issue waiting for the quickpick to
show.

sharon-wang committed:
undo changes to existing R and Python start interpreter fixtures

Create separate python-only fixture for starting and getting the selected interpreter.

sharon-wang commented:
I haven't yet implemented a fix, but here's a description of the problem that's occurring.

The error messages in the previous comment are a bit misleading. The actual issue is that the `python.setInterpreter` command doesn't show up in the CI run for some reason.

### Expected Steps
1. open quickpick
2. run command `python.setInterpreter`
3. select the interpreter based on `process.env.POSITRON_PY_VER_SEL`
4. do rest of test

### Actual Steps Occurring in CI
1. open quickpick
2. run command `python.setInterpreter`
3. can't find the command, so closes and opens the quickpick a few times to wait until the command can be found
4. still can't find the command after a few retries, so the `Python: New Python File` is clicked instead
![page@01bf1b5c39cb907e9e6d4164dba484d8-1720713130 866567](https://github.com/posit-dev/positron/assets/25834218/1a410822-4878-4e54-975f-917a568073ba)
5. the rest of the test cannot run successfully as a result

This doesn't occur with the second ipykernel test, which also runs the same code to select the interpreter. I would guess that the python extension is not quite ready at the point that the first ipykernel test is run, and so the command `python.setInterpreter` has not yet been registered for use. I'm not sure why that would be, since the default python project test would have already run, so I'd expect the IDE to be ready?

Some thoughts:
- is there a way to wait for the python extension to be ready before trying to run the command?
- why isn't this issue occurring for other tests which also try to set the interpreter at the beginning of the test?
- we could increase the retries and see if the command shows up eventually?
- desperation idea: we could add a wait to the beginning of the test and hope it works

jonvanausdeln commented on test/smoke/src/areas/positron/new-project-wizard/new-project.test.ts:
```diff
@@ -4,74 +4,155 @@
+				await pw.disabledCreateButton.isNotVisible(500);
+				await pw.nextButton.click();
+				await pw.currentOrNewWindowSelectionModal.currentWindowButton.click();
 				await app.workbench.positronExplorer.explorerProjectTitle.waitForText('myPythonProject');
```
```suggestion
				await app.workbench.positronExplorer.explorerProjectTitle.waitForText('myPythonProject');
				await app.workbench.positronConsole.waitForReady('>>>');
				await app.workbench.positronConsole.waitForConsoleContents((contents) => contents.some((line) => line.includes('restarted')));
```
I foudn that this test wasn't leaving the app in a good state. The new project was still loading, but the test finished and went on to the next test.. whcih as you pointed out, appears to not have the python extension fully loaded. This should make sure that the project from this default test if fully loaded before finishing.

Additionally in `positronConsle.waitForReady`, I would change the line that waits for the prompt to have a timeout (Github wouldln't let me add a comment to the file at the location because you hadn't changed it).. as I saw on my maching that with the above change, it would timout in `waitforReady` becuase it uses the default timout/retries.  so this change below fixed that:

```
await this.code.waitForTextContent(`${ACTIVE_CONSOLE_INSTANCE} .active-line-number`, prompt, undefined, 500)
```

sharon-wang replied:
Ah, nice find 🙌 I've added in a `waitForReady` for each of the default tests and modified `waitForReady` to accept an optional retryCount with a default of 500.

The test passed in CI in the latest run! https://github.com/posit-dev/positron/actions/runs/9907461057/job/27371205186

The Jupyter test failed because IDE startup went pretty slow in CI and it timed out. I just pushed a fix and set off a new CI run. Hope it's all green this time.

jonvanausdeln commented on test/smoke/src/areas/positron/new-project-wizard/new-project.test.ts:
```diff
@@ -4,74 +4,155 @@
+					// Start the Python interpreter and ensure ipykernel is installed
+					const interpreterInfo = await pythonFixtures.startAndGetPythonInterpreter(true);
+					expect(interpreterInfo?.path).toBeDefined();
+					const interpreterPath = interpreterInfo?.path!;
```
This ends up causing a failure for me on windows.. because `interpreterPath` is something like `~\pathtopython` but.. when the dropdown in the New Project pane has `C:\Users\JonV\pathtopython`. It's an issue with the `~` expansion in windows (or lack thereof)

sharon-wang committed:
add default retryCount = 500 to positronConsole `waitForReady`

sharon-wang committed:
call `waitForReady` for default project wizard tests

sharon-wang committed:
check the absolute interpreter path as a fallback

sharon-wang committed:
increase timeout for positronConsole.waitForReady

The Jupyter test seems to run a bit slow in CI, so we need to increase the timeout for the waitForReady call to account for slow startup.

sharon-wang committed:
remove `positronConsole.waitForReady` from R and Jupyter default tests

They are timing out in CI. They will be addressed separately.

sharon-wang committed:
add `getSelectedInterpreterInfo` to `startInterpreter`

Moves some shared types to `positronInterpreterInfo.ts` and refactors `StartInterpreter.awaitDesiredPrimaryInterpreterGroupLoaded` to be a bit simpler and also return the interpreter path.

sharon-wang committed:
get rid of bad absolute path resolution

sharon-wang committed:
support any known interpreter type in `getSelectedInterpreterInfo()`

sharon-wang committed:
cleanup `awaitDesiredPrimaryInterpreterGroupLoaded()`

sharon-wang committed:
add example comments to `selectAndGetInterpreter` in `positronConsole.ts`

sharon-wang commented on test/automation/src/positron/positronStartInterpreter.ts:
`StartInterpreter.selectInterpreter()` doesn't look to be in active use in our smoke tests. I've refactored and modified `StartInterpreter.awaitDesiredPrimaryInterpreterGroupLoaded()` to include returning the interpreter path, which may impact the functionality of `StartInterpreter.selectInterpreter()`.

Since `selectInterpreter()` isn't in active use, I'll separately add some simple smoke tests to confirm the various `StartInterpreter` methods are working and refactor as needed. I think some similar refactoring will be needed for `getSecondaryInterpreters()`, but that will also be for a separate PR.

testlabauto reviewed (dismissed):
Woohoo!

sharon-wang commented:
Just a note that your approval review was automatically changed to "stale" after I rebased on main to address some merge conflicts. I suspect this will happen with the other two Project Wizard smoke test PRs.
